### PR TITLE
feat: container management

### PR DESCRIPTION
**Pull request by m8vago**

### CODE REVIEW

File: golang/internal/grpc/grpc.go [hunk 1]
```diff
@@ -375,6 +383,15 @@ func executeDeleteContainer(ctx context.Context, req *agent.ContainerDeleteReque
 	}
 }
 
+func executeDeleteMultipleContainers(ctx context.Context, req *common.DeleteContainersRequest, deleteFn DeleteContainersFunc) {
+	log.Printf("Deleting multiple containers")
```
Comment (polaroi8d) [inline golang/internal/grpc/grpc.go:387]: Use zerolog.

File: golang/internal/grpc/grpc.go [hunk 2]
```diff
@@ -375,6 +383,15 @@ func executeDeleteContainer(ctx context.Context, req *agent.ContainerDeleteReque
...
+
+	err := deleteFn(ctx, req)
+	if err != nil {
+		log.Printf("Failed to delete multiple containers: %v", err)
```
Comment (polaroi8d) [inline golang/internal/grpc/grpc.go:391]: Use zerolog.

File: golang/internal/grpc/grpc.go [hunk 3]
```diff
@@ -502,6 +519,15 @@ func executeClose(ctx context.Context, command *agent.CloseConnection, closeFunc
 	}
 }
 
+func executeContainerCommand(ctx context.Context, command *common.ContainerCommandRequest, containerCommandFunc ContainerCommandFunc) {
+	log.Printf("Executing container command %s on %s", command.Operation.String(), command.GetId())
```
Comment (polaroi8d) [inline golang/internal/grpc/grpc.go:523]: Use zerolog.

File: golang/internal/helper/docker/container.go [hunk 4]
```diff
@@ -27,16 +27,33 @@ func DeleteContainerByName(ctx context.Context, dog *dogger.DeploymentLogger, na
 	return deleteContainerByState(ctx, dog, nameFilter, matchedContainer.State, errorReport)
 }
 
-func DeleteContainerByID(ctx context.Context, dog *dogger.DeploymentLogger, nameFilter string, errorReport bool) error {
-	matchedContainer, err := GetContainerByID(ctx, nil, nameFilter, errorReport)
+func DeleteContainerByID(ctx context.Context, dog *dogger.DeploymentLogger, idFilter string, errorReport bool) error {
+	matchedContainer, err := GetContainerByID(ctx, nil, idFilter, errorReport)
 	if err != nil {
-		return fmt.Errorf("builder could not get container (%s) to remove: %s", nameFilter, err.Error())
+		return fmt.Errorf("builder could not get container (%s) to remove: %s", idFilter, err.Error())
```
Comment (polaroi8d) [inline golang/internal/helper/docker/container.go:33]: Use zerolog.

File: golang/pkg/crane/crux/deploy.go [hunk 5]
```diff
@@ -27,3 +28,9 @@ func GetSecretsList(ctx context.Context, prefix, name string) ([]string, error)
 
 	return k8s.ListSecrets(ctx, prefix, name, cfg)
 }
+
+func DeploymentCommand(ctx context.Context, command *common.ContainerCommandRequest) error {
+	// operation := command.Operation
```
Comment (polaroi8d) [inline golang/pkg/crane/crux/deploy.go:33]: Add Todo please.

File: golang/pkg/dagent/utils/docker.go [hunk 6]
```diff
@@ -608,3 +608,40 @@ func SecretList(ctx context.Context, prefix, name string) ([]string, error) {
...
+	}
+
+	if operation == common.ContainerOperation_START_CONTAINER {
+		err = cli.ContainerStart(ctx, containerID, types.ContainerStartOptions{})
+	} else if operation == common.ContainerOperation_STOP_CONTAINER {
+		err = cli.ContainerStop(ctx, containerID, nil)
+	} else if operation == common.ContainerOperation_RESTART_CONTAINER {
+		err = cli.ContainerRestart(ctx, containerID, nil)
+	} else {
+		log.Printf("Unknown container operation %d on containerID %s", operation, containerID)
```
Comment (polaroi8d) [inline golang/pkg/dagent/utils/docker.go:628]: Use zerolog.

File: web/crux-ui/src/components/nodes/node-containers-list.tsx [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    return [
+      <span>{container.name}</span>,
+      <span className="block overflow-hidden truncate">{imageName(container.imageName, container.imageTag)}</span>,
+      <ContainerStatusTag className="inline-block" state={container.state} />,
+      <span>{utcDateToLocale(container.date)}</span>,
+      !container.ports ? null : (
+        <span className="block overflow-hidden truncate">{containerPortsToString(container.ports)}</span>
+      ),
+      <div className="flex flex-wrap gap-1 justify-center items-center">
+        {/* TODO (@m8vago): get logs  */}
```
Comment (polaroi8d) [inline web/crux-ui/src/components/nodes/node-containers-list.tsx:46]: Needed?

File: golang/internal/helper/docker/container.go [hunk 8]
```diff
@@ -27,16 +27,33 @@ func DeleteContainerByName(ctx context.Context, dog *dogger.DeploymentLogger, na
...
 
-	return deleteContainerByState(ctx, dog, nameFilter, matchedContainer.State, errorReport)
+	return deleteContainerByState(ctx, dog, idFilter, matchedContainer.State, errorReport)
+}
+
+func DeletePrefix(ctx context.Context, prefix string) error {
+	containers, err := GetAllContainersByName(ctx, nil, prefix)
+
+	if err != nil {
+		return fmt.Errorf("builder could not get containers for prefix (%s) to remove: %s", prefix, err.Error())
```
Comment (nandor-magyar) [inline golang/internal/helper/docker/container.go:46]: we are not in the builder.

File: golang/pkg/dagent/utils/docker.go [hunk 9]
```diff
@@ -608,3 +608,40 @@ func SecretList(ctx context.Context, prefix, name string) ([]string, error) {
...
+	}
+
+	return err
+}
+
+func DeleteContainers(ctx context.Context, request *common.DeleteContainersRequest) error {
```
Comment (nandor-magyar) [inline golang/pkg/dagent/utils/docker.go:634]: Merging the two DeleteContainers function is not an option? Core will not call it without adding adding the container name, so no "prefix delete" will be initiated. Having two functions for almost the same thing looks weird.

**Review by nandor-magyar — CHANGES_REQUESTED** [review]
Checked the go part, added some comments, regarding log: make sure to add log levels too.

File: golang/pkg/dagent/utils/docker.go (hunk 9 above)
Comment (m8vago) [reply golang/pkg/dagent/utils/docker.go:634]: I didn't wanted to tweak with the core itself, cause i would had to modify the request. So my strat here is to get rid of the core and resolve this issue in the future.

File: web/crux-ui/src/components/nodes/node-containers-list.tsx (hunk 7 above)
Comment (m8vago) [reply web/crux-ui/src/components/nodes/node-containers-list.tsx:46]: yap

File: golang/internal/helper/docker/container.go (hunk 8 above)
Comment (m8vago) [reply golang/internal/helper/docker/container.go:46]: i've just copied this from above, probably that's faulty as well?

File: golang/internal/helper/docker/container.go [hunk 10]
```diff
@@ -102,32 +118,34 @@ func RemoveContainer(ctx context.Context, dog *dogger.DeploymentLogger, nameFilt
 }
 
 // Check the existence of containers, then return it
-func GetAllContainersByName(ctx context.Context, dog *dogger.DeploymentLogger, nameFilter string) ([]types.Container, error) {
+func GetAllContainersByName(ctx context.Context, dog *dogger.DeploymentLogger, nameFilter string) ([]*types.Container, error) {
```
Comment (robot9706) [inline golang/internal/helper/docker/container.go:121]: What's the point of returning `nil` containers? Why not just leave them out of the array?

**Review by robot9706 — CHANGES_REQUESTED** [review]

File: golang/internal/grpc/grpc.go [hunk 11]
```diff
@@ -41,21 +41,25 @@ type ConnectionParams struct {
...
-	SecretListFunc func(context.Context, string, string) ([]string, error)
-	SelfUpdateFunc func(context.Context, string, int32) error
-	CloseFunc      func(context.Context, agent.CloseReason) error
+	DeployFunc           func(context.Context, *dogger.DeploymentLogger, *v1.DeployImageRequest, *v1.VersionData) error
+	WatchFunc            func(context.Context, string) []*common.ContainerStateItem
+	DeleteFunc           func(context.Context, string, string) error
+	SecretListFunc       func(context.Context, string, string) ([]string, error)
+	SelfUpdateFunc       func(context.Context, string, int32) error
+	CloseFunc            func(context.Context, agent.CloseReason) error
+	ContainerCommandFunc func(context.Context, *common.ContainerCommandRequest) error
```
Comment (nandor-magyar) [inline golang/internal/grpc/grpc.go:50]: no other functions have params from the proto file, at that point a peeled version of the request should be arriving independent of transport methods.
This and the following line.

File: golang/internal/helper/docker/container.go [hunk 12]
```diff
@@ -27,16 +27,32 @@ func DeleteContainerByName(ctx context.Context, dog *dogger.DeploymentLogger, na
...
+func DeletePrefix(ctx context.Context, prefix string) error {
+	containers, err := GetAllContainersByName(ctx, nil, prefix)
+	if err != nil {
+		return fmt.Errorf("could not get containers for prefix (%s) to remove: %s", prefix, err.Error())
+	}
+
+	for _, container := range containers {
+		containerDeleteErr := deleteContainerByState(ctx, nil, container.ID, container.State, false)
+		if err == nil {
+			err = containerDeleteErr
```
[the remark this answers is not in the document]
Comment (m8vago) [reply golang/internal/helper/docker/container.go:51]: sure

File: golang/internal/helper/docker/container.go (hunk 10 above)
Comment (m8vago) [reply golang/internal/helper/docker/container.go:121]: This is an optimization for the foreaches

File: golang/internal/grpc/grpc.go (hunk 11 above)
Comment (m8vago) [reply golang/internal/grpc/grpc.go:50]: noted. made a techdept ticket for it

**Review by nandor-magyar — APPROVED** [review]

**Pull request merged**